### ISSUE TITLE
multiple hostnames and `TLS_FLAVOR=mail`

### ISSUE THREAD

**Issue by patzm** [Mailu/Mailu#2463]
Hi guys,
I have a mail server running behind a Traefik reverse Proxy. The entire setup is [here](). I thus use `TLS_FLAVOR=mail`: https://github.com/patzm/dockerfiles/blob/47fb4f0351358509453f088154013c9d3baf8f99/mailu/.env#L30

I now want to host a second domain (through `HOSTNAMES`, [link](https://github.com/patzm/dockerfiles/blob/47fb4f0351358509453f088154013c9d3baf8f99/mailu/.env#L24)). Sending emails works all nice and dandy. I was able to configure all security features and DNS settings through custom records and by setting the `A` and `AAAA` records. However, email clients such as Thunderbird ⚡🐦 still complain when configuring the second hostname to fetch emails from. Specifically, my main domain (and first hostname) is `patz.family`. The second one is `patz.app`. However, connecting to `imap.patz.app:993` for instance, it returns the certificate of `patz.family`. That's kind of obvious, because the cert-dumper only dumps that certificate. [Link](https://github.com/patzm/dockerfiles/blob/47fb4f0351358509453f088154013c9d3baf8f99/mailu/docker-compose.yml#L14-L24) to Docker compose of cert-dumper. Adding a second dumper would be trivial.

**Question**: but how would I actually mount the certificate file now? The certdumper can already dump the certifications into subfolders for the domains. But could the Mailu frontend read that format too?

**Comment by Diman0** [comment]
Mailu (or rather nginx) requires a single certificate. From your description it seems traefik requests one separate certificate per domain. This does not work with Mailu (nginx). Nginx requires a single certificate with all domains/hostnames served by the web server.

 The subject name is for the first hostname in HOSTNAMES. All the other hostnames are added as subject alternative names in the certificate.

To see an example how to request such a certificate from lets encrypt,  you can refer to the Mailu code for using certbot:
https://github.com/Mailu/Mailu/blob/b5e7cad2d3da8a6581296efdd8550b8c73f819e5/core/nginx/letsencrypt.py#L12

**Comment by patzm** [comment]
Hm ok thanks for the quick response. In my understanding, this works for SANS, i.e. sub-domains like
* `mail.patz.family`
* `mail2.patz.family`
* `something.patz.family`

But I have different TLD (top-level-domains): `patz.family` and `patz.app`. How would that work? 😕

I am considering setting the `CNAME` of `patz.family` to point to `patz.app` and only generate the certificate for that. Not sure though.

**Comment by nextgens** [comment]
Subject Alternative Names in certificates don't require sharing the same TLD

**Comment by sholl** [comment]
@patz
I set the following traefik-labels inside my front-container
```
...
- "traefik.http.routers.front.tls.domains[0].main=mail.tld1.de"
- "traefik.http.routers.front.tls.domains[0].sans=imap.tld1.de, smtp.tld1.de"
- "traefik.http.routers.front.tls.domains[1].main=mail.another-tld.net"
- "traefik.http.routers.front.tls.domains[1].sans=imap.another-tld.net, smtp.another-tld.net"
...
```
No clients on both domains complain about the "wrong" certificate.

**Comment by patzm** [comment]
But are you connecting your IMAP clients (e.g. Thunderbird) also to `imap.another-tld.net` or `smtp.another-tld.net` or do all connect to `*.tld1.de`?

My configuration ([link](https://github.com/patzm/dockerfiles/blob/47fb4f0351358509453f088154013c9d3baf8f99/mailu/docker-compose.yml#L49-L61)) is very similar to yours, but differs in that I set up a second _router_. You only added more domains. Should that alone solve it?

Also: I added the missing link to _here_ in my original message in https://github.com/Mailu/Mailu/issues/2463#issue-1396903030.

**Comment by sholl** [comment]
indeed, I did not look at your config close enough. AFAIK the clients connect to *.tld1.de, but need to recheck.

**Comment by patzm** [comment]
ok, please recheck 🙏, because that's really the difference for me. If all clients connect to `*.tld1.de`, then everything works for me too. But I'd like them to be able to connect to the other TLDs as well.

**Comment by sholl** [comment]
@patzm rechecked, it is the same as you described. If connecting to imap.another-tld.net a certerror raises. Only the first certificate is used. So same behaviour as yours.

As a workarround I can think of pasting all certs from certdumper into one file and provide that file to mailu (untested though).

**Comment by patzm** [comment]
I think that SANS with other TLDs don't work fundamentally. But I can try.
But I'd think that a combination of modifying the certdumper to dump multiple certificates and configuring Mailu to read multiple certificates, and then present the appropriate certificate, depending on which domain is called. Hence this issue's title, _multiple hostnames_.

**Comment by patzm** [comment]
I didn't find time to try it yet. I am pretty sure though that a SANS for another TLD doesn't work. I would like to keep this issues open as a feature request. Can you disable the bot?
